lib/block-supports/layout.php

```diff
@@ -136,29 +141,35 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 			 * by custom css.
 			 */
 			if ( ! empty( $layout['justifyContent'] ) && array_key_exists( $layout['justifyContent'], $justify_content_options ) ) {
-				$style .= "$selector {";
-				$style .= "justify-content: {$justify_content_options[ $layout['justifyContent'] ]};";
-				$style .= '}';
+				$layout_styles[ $selector ] = array( 'justify-content' => $justify_content_options[ $layout['justifyContent'] ] );
 			}
 
 			if ( ! empty( $layout['verticalAlignment'] ) && array_key_exists( $layout['verticalAlignment'], $vertical_alignment_options ) ) {
-				$style .= "$selector {";
-				$style .= "align-items: {$vertical_alignment_options[ $layout['verticalAlignment'] ]};";
-				$style .= '}';
+				$layout_styles[ $selector ] = array( 'align-items' => $vertical_alignment_options[ $layout['verticalAlignment'] ] );
 			}
 		} else {
-			$style .= "$selector {";
-			$style .= 'flex-direction: column;';
+			$layout_styles[ $selector ] = array(
+				'flex-direction' => 'column',
+				'align-items'    => 'flex-start',
+			);
 			if ( ! empty( $layout['justifyContent'] ) && array_key_exists( $layout['justifyContent'], $justify_content_options ) ) {
-				$style .= "align-items: {$justify_content_options[ $layout['justifyContent'] ]};";
+				$layout_styles[ $selector ] = array( 'align-items' => $justify_content_options[ $layout['justifyContent'] ] );
 			} else {
-				$style .= 'align-items: flex-start;';
+				$layout_styles[ $selector ] = array( 'align-items' => 'flex-start' );
 			}
-			$style .= '}';
 		}
 	}
 
-	return $style;
+	if ( ! empty( $layout_styles ) ) {
+		// Add to the style engine store to enqueue and render layout styles.
+		gutenberg_style_engine_add_to_store( 'layout-block-supports', $layout_styles );
+
+		// Return compiled layout styles to retain backwards compatibility.
+		// Since https://github.com/WordPress/gutenberg/pull/42452 we no longer call wp_enqueue_block_support_styles in this block supports file.
+		return gutenberg_style_engine_get_stylesheet( 'layout-block-supports' );
```

Comment:
Oh, this is a tricky one — the return here is cumulative in that the output appears to contain all previously registered layout styles for that store.

If I log out this return value, here's how it looks in console output (note that the first line is a single rule, and the subsequent lines are that rule plus the next instance of the rendering):

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/14988353/181427813-4f10a942-2e01-4dc7-89b9-315394017c5f.png">

Reply:
> Oh, this is a tricky one — the return here is cumulative in that the output appears to contain all previously registered layout styles for that store.

Yeah, I think we might have to bite the bullet and manually output them _in situ_

Reply:
In my alternative PR I had added a `$cleanup_rules` in the `get_css()` method to avoid this issue. See https://github.com/WordPress/gutenberg/pull/42592/files#diff-f23bfb067de6325fe30b9ec68f807342a3286e40b87937bcb806ea285e0252a3R44
Basically whenever CSS was getting printed by the `WP_Style_Engine_Processor` class, I had the option to pass `true` as an arg to the method and "reset" the stores so there's no duplication of styles.

Reply:
> Basically whenever CSS was getting printed by the WP_Style_Engine_Processor class, I had the option to pass true as an arg to the method and "reset" the stores so there's no duplication of styles.

Interesting, thanks! 

Sounds like it could help us in this situation: we only want to return the styles that are collected in the body of the layout function during a single execution. What we do for backward compatibility! 😄 

I haven't tested so could be way wrong here, but I think the snag is that nothing is rendered yet.  We're relying on the full contents of the store existing later when enqueuing.







What we do for backwards compatibility! 😄